src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs

```diff
@@ -101,12 +102,29 @@ IEnumerator IEnumerable.GetEnumerator()
 
         void ICollection.CopyTo(Array array, int arrayIndex)
         {
+            CheckCopyToArguments(array, arrayIndex);
             foreach (var item in this)
             {
                 array.SetValue(new DictionaryEntry(item.Key, item.Value), arrayIndex++);
             }
         }
 
+        private void CheckCopyToArguments(Array array, int arrayIndex)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
```

Comment:
nit: We should eventually use `ErrorUtilities` here. Something like `ErrorUtilities.VerifyThrowArgumentNull` and some equivalent for the other checks if possible.

Reply:
What are the rules around when we should use ErrorUtilities vs. not? ErrorUtilities when doing something MSBuild-specific, but I would've thought that in making a DS, we should try to match the real one as closely as possible.

Reply:
I'm not sure, and the summary of the `ErrorUtilities` class doesn't say much: `This class contains methods that are useful for error checking and validation.` My assumption was that we should generally use those methods. I don't think using them hurts in this case, but as long as we're _specifically choosing_ to match the inherited class then I've no problem with that.

Reply:
I think in this case using the bare `throw new ...` is OK as we're not exposing the data structure publicly and want to match the behavior of the real .NET 4.5 implementation.